src/kaleidoscope_internal/LEDModeManager.h

```diff
@@ -0,0 +1,428 @@
+/* Kaleidoscope - Firmware for computer input devices
+ * Copyright (C) 2013-2018  Keyboard.io, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "kaleidoscope_internal/array_like_storage.h"
+#include "kaleidoscope/plugin.h"
+#include "kaleidoscope/plugin/LEDMode.h"
+
+#include <stddef.h>
+
+// To enable placement new, we need to supply a global operator
+// function.
+//
+inline void* operator new (size_t, void* __p) throw() {
+  return __p;
+}
+
+namespace kaleidoscope {
+
+namespace plugin {
+// Forward declarations to avoid header inclusions
+class LEDControl;
+class LEDModeInterface;
+} // end namespace plugin
+
+namespace internal {
+namespace led_mode_management {
+
+// Functions of this type allocate a new LED mode within a global
+// buffer and give it access to it's parent plugin.
+// This allows a dynamic LED mode to access member data of its parent
+// plugin, which can help to reduce its RAM footprint.
+// Some dynamic LED modes might ignore the parent pointer entirely, e.g. if
+// they do not have a configurable internal state.
+//
+// We use a factory function instead of a polymorphic
+// class to generate LED modes. By this means we can avoid the additional
+// cost of vtables which safes some program memory.
+// Conceptually, the factory function approach is similar to having
+// a polymorphic class with only one method, that is accessed directly
+// instead of the detour via the vtable.
+//
+typedef kaleidoscope::plugin::LEDMode*
+(*LEDModeFactoryFunc)(void *raw_buffer,
+                      void *parent,
+                      uint8_t mode_id
+                     );
+
+// A templated implementation of LEDModeFactoryFunc.
+//
+template<typename ParentPlugin__>
+static kaleidoscope::plugin::LEDMode *
+generateLEDMode(void *raw_buffer,
+                void *parent_plugin,
+                uint8_t mode_id
+               ) {
+  // We know the type of the parent plugin via the template parameter
+  // ParentPlugin__, thus it is safe to cast to the actual type.
+  //
+  auto parent_plugin_actual = static_cast<ParentPlugin__*>(parent_plugin);
```

Comment:
I'm guessing that the reason `parent_plugin` is a generic `void*` pointer is because the caller is reading it from PROGMEM, and doesn't bother to do the conversion before calling `generateLEDMode()`. Is that right?

Reply:
No, it's because `LEDModeManager` is designed entirely type agnostic and thus generic with respect to plugins and LED modes. Only the factory functions know the actual type of the plugins they operate on. 

This mechanism would work the same if every plugin that deals with LED modes would have its own static `LEDModeFactoryFunc` implementation. Those pointers could as well be stored in the `LEDModeFactory`s. 

The reason why I went with the template function based approach is that I wanted to spare the plugin developer to implement such functions or add some magic macros that would implement them. Templates are just easier to handle in terms of transparency and compiler errors than magic macros.

Reply:
If the caller already knows what `ParentPlugin__` is (maybe it would be better to call it `ParentPluginType__`), why is `parent_plugin` of type `void*` instead of `ParentPlugin__*`?

Reply:
> If the caller already knows what ParentPlugin__ is (maybe it would be better to call it ParentPluginType__)

I know, that's confusing. The caller does not know the type of `ParentPlugin__`. `ParentPlugin__` is a template parameter. It is only known at compile time and only in the context, where the function pointer is assigned (which is in line 202). The caller, `LEDModeManager` is completely agnostic of the plugin type. Thus, a generic implementation is possible. Whether we call it `ParentPlugin__` or `ParentPluginType__` doesn't matter. As from the template definition it is obvious that it is a type, not a variable, I tend to skip the suffix `Type`.

> why is parent_plugin of type void* instead of ParentPlugin__*?

Same answer as above, because the calling context does not know the actual type. We could change the `void *parent_plugin` to something less frightening than a void ptr, though. I will think about that and push a change with some more comments, later.